firmware/library/L3_HAL/apds_9960.hpp

```diff
@@ -1,20 +1,444 @@
+
+    SetMode(0, 1);  // POWERON
+    SetMode(3, 1);  // WAIT
+    SetMode(2, 1);  // PROXIMITY
+    SetMode(6, 1);  // GESTURE
+    // kENABLE = 0100'1101
+  }
+  bool DisableGesture() override
+  {
+    if (!SetMode(6, 0)) return false;
+    return true;
+  }
+  uint8_t ReadGestureMode() override
+  {
+    uint8_t value = 0;
+    value = ReadReg(kGCONF4);
+    value &= ~(0xFE);
+    return value;
+  }
+  bool CheckIfGestureOccured() override
+  {
+    uint8_t gstatus = 0;
+    gstatus = ReadReg(kGSTATUS);
+
+    gstatus &= 0b00000001;
+
+    if (gstatus == 1)
+    {
+    return true;
+    } else
+    {
+    return false;
+    }
+  }
+  void ReadGestureFIFO(uint8_t * data, uint8_t fifolvl) override
+  {
+    uint8_t fifo_registers = kGFifoU;
+    i2c.WriteThenRead(kApds9960Address, &fifo_registers, 1, data, (4*fifolvl));
+  }
+  uint8_t GetGestureFIFOLvl() override
+  {
+    uint8_t value = 0;
+    value = ReadReg(kGSTATUS);
+    value &= 0b00000010;  // GFOV Mask
+    if (value)
+    {
+      uint8_t level = 0;
+      level = ReadReg(kGFLVL);
+
+      uint8_t throw_away[128];
+      while (level != 0)
+      {
+        ReadGestureFIFO(throw_away, level);
+        level = ReadReg(kGFLVL);
+      }
+    }
+    return ReadReg(kGFLVL);
+  }
+  void ProcessGestureData(uint8_t level) override
+  {
+    for (int i = 0; i < (4*level); i+=4)
+    {
+      if (abs((gfifo_data_[i] - gfifo_data_[i+1]))
+            > abs((gfifo_data_[i+2] - gfifo_data_[i+3])))
+      {
+        if ((gfifo_data_[i] - gfifo_data_[i+1]) < up_sensitivity_)  // Up
+        {
+            points_[index_] = Direction::kUp;
+            index_++;
+        } else if ((gfifo_data_[i] - gfifo_data_[i+1])
+                    > down_sensitivity_)  // Down
+        {
+            points_[index_] = Direction::kDown;
+            index_++;
+        }
+      } else
+      {
+        if ((gfifo_data_[i+2] - gfifo_data_[i+3])
+                  < left_sensitivity_)  // Left
+        {
+            points_[index_] = Direction::kLeft;
+            index_++;
+        } else if ((gfifo_data_[i+2] - gfifo_data_[i+3])
+                    > right_sensitivity_)  // Right
+        {
+            points_[index_] = Direction::kRight;
+            index_++;
+        }
+      }
+      if ((gfifo_data_[i] + gfifo_data_[i+1] + gfifo_data_[i+2]
+                  + gfifo_data_[i+3]) > far_sensitivity_ )  // Near
+      {
+          near_count_++;
+      } else if ((gfifo_data_[i] + gfifo_data_[i+1] + gfifo_data_[i+2]
+                  + gfifo_data_[i+3]) < near_sensitivity_ )  // Far
+      {
+          far_count_++;
+      } else
+      {
+          // do nothing
+      }
+    }
+  }
+  char DecodeGestureData() override
+  {
+    for (int i = 0; i < 8; i++)
+    {
+      switch (points_[i])
+      {
+      case Direction::kUp:
+          if (points_[i+1] == Direction::kDown)
+          {
+              // DEBUG_PRINT("Down Gesture\n");
+              return static_cast<char>(Gesture::kSwipeDOWN);
```

Comment:
Prefer not to use multiple return statements. Have a singular Gesture type variable above and set it based on which case has been set. Then return it at the end of the function, with the necessary cast.